scio-parquet/src/main/scala/com/spotify/scio/parquet/avro/ParquetAvroIO.scala

```diff
@@ -87,23 +91,80 @@ final case class ParquetAvroIO[T: ClassTag: Coder](path: String) extends ScioIO[
     val resource =
       FileBasedSink.convertToFileResourceIfPossible(ScioUtil.pathWithShards(path))
     val prefix = StaticValueProvider.of(resource)
-    val usedFilenamePolicy =
-      DefaultFilenamePolicy.fromStandardParameters(prefix, null, params.suffix, false)
-    val destinations = DynamicFileDestinations.constant[T](usedFilenamePolicy)
+    val fileNamePolicy =
+      if (params.windowedFilenameFunction.isDefined || params.filenameFunction.isDefined) {
+        createFilenamePolicy(
+          resource,
+          params.windowedFilenameFunction,
+          params.filenameFunction,
+          params.suffix
+        )
+      } else {
+        DefaultFilenamePolicy.fromStandardParameters(prefix, null, params.suffix, false)
+      }
+    val destinations = DynamicFileDestinations.constant[T](fileNamePolicy)
     val sink = new ParquetAvroSink[T](
       prefix,
       destinations,
       writerSchema,
       job.getConfiguration,
       params.compression
     )
-    val t = WriteFiles.to(sink).withNumShards(params.numShards)
+    val t =
+      if (data.internal.getWindowingStrategy != WindowingStrategy.globalDefault()) {
+        WriteFiles.to(sink).withNumShards(params.numShards).withWindowedWrites()
+      } else {
+        WriteFiles.to(sink).withNumShards(params.numShards)
+      }
     data.applyInternal(t)
     tap(ParquetAvroIO.ReadParam[T, T](writerSchema, null, identity))
   }
 
   override def tap(params: ReadP): Tap[T] =
     ParquetAvroTap(ScioUtil.addPartSuffix(path), params)
+
+  def createFilenamePolicy(
+    baseFileName: ResourceId,
+    windowedFilenameFunction: Option[(Int, Int, BoundedWindow, PaneInfo) => String],
+    filenameFunction: Option[(Int, Int) => String],
+    filenameSuffix: String
+  ): FilenamePolicy = {
+
+    new FileBasedSink.FilenamePolicy {
+
+      override def windowedFilename(
+        shardNumber: Int,
+        numShards: Int,
+        window: BoundedWindow,
+        paneInfo: PaneInfo,
+        outputFileHints: FileBasedSink.OutputFileHints
+      ): ResourceId = {
+        windowedFilenameFunction
```

Comment:
Minor: see if we can make `Option[F]` `F` instead to get rid of the 2 `.map` & object allocation? If either one of the functions can be absent, we can provide a fallback e.g. `(Int, Int) => throw new IllegalArgumentException()` with a meaningful message?

Reply:
So I'm currently giving the new WriteParams a default of `None`, but I think what you're proposing (change the default to a function that throws an exception instead) would get rid of all of the Options and allow us to get rid of the `getOrElse(???)` as well. I'll do this, as it'll clean things up a little more.